### ISSUE TITLE
Using a derive proc-macro from macro_rules!() results in a compile error about `self` parameter.

### ISSUE THREAD

**Issue by manuthambi** [rust-lang/rust#121558]
I tried this code:

```rust
pub trait BitwiseEq {
    fn bitwise_eq(&self, other: &Self) -> bool;
}

impl BitwiseEq for i32 {
    fn bitwise_eq(&self, other: &Self) -> bool { *self == *other }
}

// This is fine
#[derive(derive::BitwiseEq)]
pub struct Wrap1(i32);

// This is fine
#[derive(derive::BitwiseEq)]
pub struct Wrap2 { v: i32 }

macro_rules! impl_bitwise3 {
    ($name:ident, $type:ident) => {
        #[derive(derive::BitwiseEq)]
        pub struct $name { v: $type }
    }
}
// This is fine
impl_bitwise3!(Wrap3, i32);

macro_rules! impl_bitwise4 {
    ($name:ident, $type:ident) => {
        #[derive(derive::BitwiseEq)]
        pub struct $name(pub $type);
    }
}
// This is fine
impl_bitwise4!(Wrap4, i32);

macro_rules! impl_bitwise5 {
    ($name:ident, $type:ident) => {
        #[derive(derive::BitwiseEq)]
        pub struct $name($type);
    }
}
// This produces a compile error
impl_bitwise5!(Wrap5, i32);
```

The proc-macro is this
```rust
extern crate proc_macro;
use proc_macro2::{Span, TokenStream};
use quote::{quote, quote_spanned};
use syn::spanned::Spanned;
use syn::{parse_macro_input, Data, DeriveInput, Error, Fields, Index, Result};

#[proc_macro_derive(BitwiseEq)]
pub fn bitwise_eq_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    match derive(input) {
        Ok(v) => v.into(),
        Err(e) => e.to_compile_error().into(),
    }
}

fn derive(input: DeriveInput) -> Result<TokenStream> {
    // Used in the quasi-quotation below as `#name`.
    let name = input.ident;

    // Generate an expression to compare each field.
    let eq_expr = match &input.data {
        Data::Struct(data) => {
            match &data.fields {
                Fields::Named(fields) => {
                    // Expands to an expression like
                    //
                    //     true && BitwiseEq::bitwise_eq(&self.x, &other.x) && BitwiseEq::bitwise_eq(&self.y, &other.y) && ...
                    //
                    // We take some care to use the span of each `syn::Field` as the span of the corresponding `bitwise_eq`
                    // call. This way if one of the field types does not implement `BitwiseEq` then the compiler's error message
                    // underlines which field it is. An example is shown in the readme of the parent directory.
                    let recurse = fields.named.iter().map(|f| {
                        let name = &f.ident;
                        quote_spanned! {f.span()=> crate::BitwiseEq::bitwise_eq(&self.#name, &other.#name) }
                    });
                    quote! {
                        true #(&& #recurse)*
                    }
                }
                Fields::Unnamed(fields) => {
                    // Expands to an expression like
                    //
                    //     true && BitwiseEq::bitwise_eq(&self.0, &other.0) && BitwiseEq::bitwise_eq(&self.1, &other.1) && ...
                    let recurse = fields.unnamed.iter().enumerate().map(|(i, f)| {
                        let index = Index::from(i);
                        quote_spanned! {f.span()=> crate::BitwiseEq::bitwise_eq(&self.#index, &other.#index) }
                    });
                    quote! {
                        true #(&& #recurse)*
                    }
                },
                Fields::Unit => {
                    // Unit structs are always equal to each other
                    quote!(true)
                }
            }
        }
        Data::Enum(_) | Data::Union(_) => {
            return Err(Error::new(Span::call_site(), "`#[derive(BitwiseEq)]` not implemented for `enums`s or `union`s."));
        }
    };

    // The generated impl.
    let expanded = quote! {
        impl crate::BitwiseEq for #name  {
            #[inline]
            fn bitwise_eq(&self, other: &Self) -> bool {
                #eq_expr
            }
        }
    };
    Ok(expanded)
}
```

It should have compiled. But I get an compile error. The compile error only happens in `impl_bitwise5!` which uses a tuple struct where the inner value is not `pub`.

```
cargo check
    Checking macrobug v0.0.0 (/home/manu/programs/rust/macro-bug)
error[E0424]: expected value, found module `self`
  --> src/lib.rs:42:23
   |
37 |         #[derive(derive::BitwiseEq)]
   |                  ----------------- this function has a `self` parameter, but a macro invocation can only access identifiers it receives from parameters
...
42 | impl_bitwise5!(Wrap5, i32);
   |                       ^^^ `self` value is a keyword only available in methods with a `self` parameter

error[E0425]: cannot find value `other` in this scope
  --> src/lib.rs:42:23
   |
42 | impl_bitwise5!(Wrap5, i32);
   |                       ^^^ not found in this scope
```

### Meta

The same error happens in both `stable` and `nightly`

`rustc --version --verbose`:
```
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: x86_64-unknown-linux-gnu
release: 1.76.0
LLVM version: 17.0.6
```

I looked at the output generated by the derive macro, using `cargo expand`, and it looks fine.
```rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
pub trait BitwiseEq {
    fn bitwise_eq(&self, other: &Self) -> bool;
}
impl BitwiseEq for i32 {
    fn bitwise_eq(&self, other: &Self) -> bool {
        *self == *other
    }
}
pub struct Wrap1(i32);
impl crate::BitwiseEq for Wrap1 {
    #[inline]
    fn bitwise_eq(&self, other: &Self) -> bool {
        true && crate::BitwiseEq::bitwise_eq(&self.0, &other.0)
    }
}
pub struct Wrap2 {
    v: i32,
}
impl crate::BitwiseEq for Wrap2 {
    #[inline]
    fn bitwise_eq(&self, other: &Self) -> bool {
        true && crate::BitwiseEq::bitwise_eq(&self.v, &other.v)
    }
}
pub struct Wrap3 {
    v: i32,
}
impl crate::BitwiseEq for Wrap3 {
    #[inline]
    fn bitwise_eq(&self, other: &Self) -> bool {
        true && crate::BitwiseEq::bitwise_eq(&self.v, &other.v)
    }
}
pub struct Wrap4(pub i32);
impl crate::BitwiseEq for Wrap4 {
    #[inline]
    fn bitwise_eq(&self, other: &Self) -> bool {
        true && crate::BitwiseEq::bitwise_eq(&self.0, &other.0)
    }
}
pub struct Wrap5(i32);
impl crate::BitwiseEq for Wrap5 {
    #[inline]
    fn bitwise_eq(&self, other: &Self) -> bool {
        true && crate::BitwiseEq::bitwise_eq(&self.0, &other.0)
    }
}
```
</details>

**Comment by manuthambi** [comment]
I have a minimal example. How should I upload it?

**Comment by saethlin** [comment]
I prefer code blocks in issue comments like you did above. Even if it's multiple files, then at least it doesn't rely on a commit staying around in someone's repo.

**Comment by manuthambi** [comment]
While trying to simplify the proc-macro, I think I figured out what is going on. The issue is the following line in the proc-macro above: 
```rust
quote_spanned! {f.span()=> crate::BitwiseEq::bitwise_eq(&self.#index, &other.#index) }
``` 

When I changed `quote_spanned!` to `quote!`, it works. The `f.span()` refers to the `i32` token in `impl_bitwise5!(Wrap5, i32);`.   When he compiler sees the `self` in the expanded code, and the span for which is outside the `macro_rules`, it seems to be looking for `self` outside the `macro_rules!` call-site and errors out.  

I don't fully understand how spans interact with macro hygiene, but it makes sense for the compiler to show some kind of error. But the error was very confusing. 

Anyways, I know how to fix my code.  I'll close this issue.

Thanks for looking into it!
